lib/puppet/pops/types/type_calculator.rb

```diff
@@ -1192,27 +1200,44 @@ def tuple_entry_at(tuple_t, from, to, index)
   # @api private
   #
   def assignable_PStructType(t, t2)
-    return true if t == t2 || t.elements.empty? && (t2.is_a?(Types::PHashType))
-    h = t.hashed_elements
+    elements = t.elements
     if t2.is_a?(Types::PStructType)
       h2 = t2.hashed_elements
-      h.size == h2.size && h.all? {|k, v| assignable?(v, h2[k]) }
+      match_count = 0
+      ok = elements.reduce(true) do |memo, e|
+        break false unless memo
+        t2e = h2[e.name]
+        if t2e.nil?
+          is_type_optional?(e.type)
+        else
+          match_count += 1
+          assignable?(e.type, t2e)
+        end
+      end
+      ok && match_count == h2.size
     elsif t2.is_a?(Types::PHashType)
       size_t2 = t2.size_type || @collection_default_size_t
       size_t = Types::PIntegerType.new
-      size_t.from = size_t.to = h.size
+      size_t.from = elements.reduce(0) { |c, e| is_type_optional?(e.type) ? c : c + 1 }
+      size_t.to = elements.size
       # compatible size
       # hash key type must be string of min 1 size
       # hash value t must be assignable to each key
       element_type = t2.element_type
       assignable_PIntegerType(size_t, size_t2) &&
-        assignable?(@non_empty_string_t, t2.key_type) &&
-        h.all? {|k,v| assignable?(v, element_type) }
+        (size_t2.to == 0 || assignable?(@non_empty_string_t, t2.key_type)) &&
+          elements.all? {|e| assignable?(e.type, element_type) }
     else
       false
     end
   end
 
+  # api private
+  #
+  def is_type_optional?(t)
```

Comment:
Data and Any also accepts nil. Not a fan of special methods for checking type. We have one or so already. I would rather it went in the other direction with fewer ways to check if something has a particular type.